packages/expect/src/types.ts

```diff
@@ -118,32 +118,39 @@ export interface AsymmetricMatchers {
   stringMatching(sample: string | RegExp): AsymmetricMatcher;
 }
 
-type PromiseMatchers = {
+// if T is a function, return its parameters array type, otherwise return an unknown array type
```

Comment:
Hm.. In this use case, can it be not a function? I mean, isn’t it that `toHaveBeenCalledWith` should be used with functions. Or?

I just look at `toMatchSnapshot` again thinking that constraining types for the matcher would make sense here too. Do I miss something?

https://github.com/facebook/jest/blob/e703e6ed1861595d56e25a764642da5a32b32c5a/packages/jest-snapshot/src/types.ts#L38-L39

Reply:
@mrazauskas 
Yes, `toHaveBeenCalledWith` should only be called on functions, but we don't have a way to constrain it at that level (because the `Matcher` also handles non-function matchers. I tried doing something on [this](https://github.com/royhadad/jest/compare/main...royhadad:jest:type_extend_different_for_function_and_not_function?expand=1#diff-810c1bb9ffac48cfe14f7dbbaf10fdf5725e06feecb99c657d49f3c7712bcd00R363) branch but it is not really possible.

So to sum up, yes, in this use case it can still not be a function (and will likely fail the test at run time)